### ISSUE TITLE
Consider implementing soft-timeouts using signals

### ISSUE THREAD

**Issue by lerks** [cms-dev/cms#851]
CMS may be used in some optimization contests where submissions have an allotted time to get the best (approximate?) solution possible. In that case it's useful to notify the submission when its evaluation time is about to expire so that it can stop the computation and output its result. A neat, robust (and common?) way to do it is by sending a signal to the program (SIGINT or SIGTERM) a short time before the timeout is reached. We should investigate whether it makes sense for CMS to provide such a mechanism and how to best implement it.

**Comment by andreyv** [comment]
The `SIGXCPU` signal actually implements this. The kernel sends this signal when the soft CPU limit is reached, and the process can catch the signal.

**Comment by gollux** [comment]
Wouldn't it be better to provide the submission with the value of the time limit? This way, optimization algorithms can not only terminate when the timeout gets close, but also adjust parameters of the algorithm (e.g., population size in an evolutionary algorithm).

**Comment by gollux** [comment]
Startup time can be accounted for quite easily: just ask the OS to tell you
how much did your process already spend, for example by the times() syscall.

To make it easier for the contestants, you can provide them with a library
which can return the time limit and the time already spent.

**Comment by CristianCantoro** [comment]
So, I went ahead and coded up a solution that I think may answer my needs and maybe give a perspective on what it is said above, I put the code in this [gist][1]. I need some guidance to understand how to make it work with CMS. 

Recap of my problem: we want to give students a task for which they have to provide an approximated answer, so the more time they compute the more precise an answer they should get, but they will have a time limit. So they need to compute their answer and print it, we imagine they can compute the answer iteratively, with each iteration giving a more precise answer that the previous ones.

In the [gist][1], you will find 3 files:
* `helpers.hpp`, with the function definitions
* `helpers.cpp`, with the implementation
* `test_signal.cpp` this would be a sample submission (in this case it does nothing, just throws random numbers around and waits for the timeout.

The idea is to provide contestants with the helpers library (`helpers.hpp` and `helpers.cpp`), so they can use it for their local tests.

What they need to do is to include `helpers.hpp` and put the a call to `helpers::setup()` at the beginning of their `main`:
```cpp
#include "helpers.hpp"
// additional headers or code as needed

int main() {
    helpers::setup();
   // their algorithm
}
```

`setup()` does two things: 
* calls `set_handler()`, which sets the signal handler function for `SIGXCPU` using `sigaction`, and
* calls `set_limit()`, which sets the soft limit on CPU time `RLIMIT_CPU` to 2 seconds. 

The handler simply catches `SIGXCPU` and exits the program.

The library also provides the following functions:
*   `get_start_time()`, returns a `std::chrono::system_clock::time_point` object containing the timestamp of when the program was started;
*  `get_elapsed_time()`, returns the difference between the starting time and the current time in microseconds;
* `has_reached_timeout()`, returns a boolean which is `true` if the soft limit has been reached, `false` otherwise.

The idea is that contestants should compute their best value and print them as soon as they have computed them. When the soft limit is reached the program exits, closing the output file. The actual time limit of the task on CMS would be a little more than the soft time limit, say `2.1` seconds when the soft time limit is 2 seconds.

[1]: https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d

**Comment by lerks** [comment]
First, there is one thing I don't understand. Your handler for `SIGXCPU` causes the whole program to exit immediately after you set the internal timeout flag to true. How will the contestant code be able to see that the value of that flag has changed and act accordingly?

Anyway, the usual pattern when using a grader in CMS is that the main function is implemented by the grader, which does the necessary setup and then calls a function whose implementation is provided in the contestant code. Your pattern is different but right now I don't see any reason it shouldn't work.

To have CMS build the submitted contestant code together with your grader you need to use the Batch task type, change its first parameter to "Submissions are compiled with a grader", add both the `.cpp` and the `.hpp` files to the task as managers and call them `grader.cpp` and `grader.hpp` respectively. You can make sure that this is set up correctly by going on the task's page on CWS and looking at the compilation command: it should include `grader.cpp` together with the contestant code.

**Comment by andreyv** [comment]
Here are some comments in addition to @lerks's that could help improve the code:

- The signal handler doesn't need to be `extern "C"`. The linkage shouldn't matter, the function only needs to match the `void (*)(int)` signature and could be `static`.
- The signal flag should have type [`volatile std::sig_atomic_t`](http://en.cppreference.com/w/cpp/utility/program/sig_atomic_t). The volatile part is especially important.
- Looks like the two mechanisms, the clock and the signal, are redundant. `has_reached_timeout()` could be implemented in terms of `get_elapsed_time()`.
- `std::chrono::high_resolution_clock` returns wall time, not user time, so it shouldn't be used to measure CPU time. I don't think there is a user time clock in `std::chrono` now. You should use `std::clock()` or `clock_gettime(CLOCK_PROCESS_CPUTIME_ID, …)`

**Comment by CristianCantoro** [comment]
@lerks said:
> First, there is one thing I don't understand. Your handler for `SIGXCPU` causes the whole program to exit immediately after you set the internal timeout flag to true. How will the contestant code be able to see that the value of that flag has changed and act accordingly?

@andreyv said:
> Looks like the two mechanisms, the clock and the signal, are redundant. `has_reached_timeout()` could be implemented in terms of `get_elapsed_time()`.

I started by implementing  `has_reached_timeout()` and then thought that I preferred the signal handler to exit immediately. I am keeping   `has_reached_timeout()` for the moment in case we change our mind on it, but I like the idea of exiting on signal much more.

@lerks thanks for the instructions, I will try them.
@andreyv thanks for the feedback, I will implement it now.

**Comment by CristianCantoro** [comment]
@andreyv, I don't get what are you referring with:
> The signal flag should have type [`volatile std::sig_atomic_t`](http://en.cppreference.com/w/cpp/utility/program/sig_atomic_t). The volatile part is especially important.

Can you clarify, please?

**Comment by lerks** [comment]
Having the signal handler (installed by the admin-provided grader) kill the
program is no different (from the contestant's code's perspective) than CMS
killing the evaluation when the time limit is exceeded. Hence I really
don't see the need to have that. As Andrey said all could just be based off
the clock instead.

**Comment by CristianCantoro** [comment]
@lerks said:
> Having the signal handler (installed by the admin-provided grader) kill the program is no different (from the contestant's code's perspective) than CMS killing the evaluation when the time limit is exceeded. Hence I really don't see the need to have that. As Andrey said all could just be based off the clock instead.

Oh, I see what you are saying. I agree in general, the point is that I want to be open for the moment to both the ideas so that we can experiment a little and then decide.

We need the signal handler exiting the program because in that way the program finishes before CMS time limit is reached, which cause the fact hat the submission gets 0 points regardless, the testing does not show the output, etc. (t

**Comment by CristianCantoro** [comment]
I was able to add the files to a tast called `test`, I had rename them as `helpers.cpp -> graders.cpp` and `helpers.hpp -> test.h`.

It also seems to be working as far as picking up the files when compiling.

The problem is that it seems that the `SIGXCPU` gets thrown after around 3 seconds. I don't know why.

ny help to troubleshoot the problem?

**Comment by lerks** [comment]
The signal is raised after your program consumes two seconds "worth" of computation time. In other terms, after it uses as many CPU cycles as if it were the *only* task running on (a single core of) your CPU for two seconds, utilizing 100% of its computational capacity. Since in general the CPU is shared with other tasks the program most probably received a capacity share that is smaller than 100%. Hence it took an effective three seconds to receive the two seconds "worth" of power.

As Andrey pointed out above, that is the reason you should use clock over chrono: the latter reports the effective external (so-called wall-clock) time, whereas clock reports how many seconds worth of computation time you've used so far. (To be redundant: when SIGXCPU is raised, clock will return exactly 2, whereas chrono will probably return something greater than 2).

Also be aware that (if I remember correctly) a task configured with a time limit of T seconds means that the program receives T seconds worth of computation time but also that it is killed after 2T+1 seconds of wall-clock time, to prevent getting stuck in the case of task caught in a deadlock. Hence if your system is too loaded your submissions may be hitting the wall-clock time limit before fully using the computational capacity they have at their disposal.

**Comment by CristianCantoro** [comment]
@lerks said:
> As Andrey pointed out above, that is the reason you should use clock over chrono: the latter reports the effective external (so-called wall-clock) time, whereas clock reports how many seconds worth of computation time you've used so far. (To be redundant: when SIGXCPU is raised, clock will return exactly 2, whereas chrono will probably return something greater than 2).

I am using `clock_gettime` (see the [gist][1] and I am testing the submission on a virtual machine that is running just CMS.

If I run the program as compiled in the sandbox with `/usr/bin/time`, this is what I get:
```
$ /usr/bin/time ./test
i: 0  - nmsec: 10 - program has been running for 113753 microseconds
i: 1  - nmsec: 17 - program has been running for 201722 microseconds
i: 2  - nmsec: 18 - program has been running for 289233 microseconds
i: 3  - nmsec: 17 - program has been running for 377377 microseconds
i: 4  - nmsec: 8  - program has been running for 465339 microseconds
i: 5  - nmsec: 5  - program has been running for 553103 microseconds
i: 6  - nmsec: 13 - program has been running for 642041 microseconds
i: 7  - nmsec: 12 - program has been running for 729865 microseconds
i: 8  - nmsec: 20 - program has been running for 817914 microseconds
i: 9  - nmsec: 7  - program has been running for 911163 microseconds
i: 10 - nmsec: 5  - program has been running for 1002374 microseconds
i: 11 - nmsec: 20 - program has been running for 1098314 microseconds
i: 12 - nmsec: 20 - program has been running for 1195728 microseconds
i: 13 - nmsec: 19 - program has been running for 1286216 microseconds
i: 14 - nmsec: 12 - program has been running for 1404345 microseconds
i: 15 - nmsec: 18 - program has been running for 1498474 microseconds
i: 16 - nmsec: 10 - program has been running for 1587097 microseconds
i: 17 - nmsec: 8  - program has been running for 1674771 microseconds
i: 18 - nmsec: 17 - program has been running for 1764269 microseconds
i: 19 - nmsec: 9  - program has been running for 1853130 microseconds
i: 20 - nmsec: 16 - program has been running for 1940765 microseconds
i: 21 - nmsec: 5  - program has been running for 2029493 microseconds
i: 22 - nmsec: 19 - program has been running for 2119086 microseconds
i: 23 - nmsec: 8  - program has been running for 2207212 microseconds
i: 24 - nmsec: 7  - program has been running for 2296431 microseconds
i: 25 - nmsec: 12 - program has been running for 2384744 microseconds
i: 26 - nmsec: 14 - program has been running for 2473549 microseconds
i: 27 - nmsec: 7  - program has been running for 2563899 microseconds
i: 28 - nmsec: 20 - program has been running for 2666980 microseconds
i: 29 - nmsec: 6  - program has been running for 2779558 microseconds
i: 30 - nmsec: 9  - program has been running for 2890687 microseconds
i: 31 - nmsec: 13 - program has been running for 2997462 microseconds
i: 32 - nmsec: 8  - program has been running for 3115847 microseconds
3.08user 0.06system 0:03.16elapsed 99%CPU (0avgtext+0avgdata 1636maxresident)k
0inputs+8outputs (0major+57minor)pagefaults 0swaps
```
I did not experience this behavior at all while testing on my machine (which is running lots of other stuff, included the browser, etc.).

So, what I am missing?


[1]: https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d

**Comment by lerks** [comment]
I don't really know `clock_gettime`. Why did you choose to use that over `clock`[1]?

[1] http://www.cplusplus.com/reference/ctime/clock/

**Comment by lerks** [comment]
As I'm more familiar with Python that with C++, I tried implementing some proof-of-concept and got this, which should be doing almost the same thing you are doing:
```py
#!/usr/bin/python3

import resource
import signal
import time

timed_out = False

def on_timeout(signum, frame):
    global timed_out
    timed_out = True

signal.signal(signal.SIGXCPU, on_timeout)

soft, hard = resource.getrlimit(resource.RLIMIT_CPU)
resource.setrlimit(resource.RLIMIT_CPU, (1, hard))

t = 0
while not timed_out:
    for _ in range(100_000):
        t += 1
        t *= 2
        t %= 1_000_000_009
    time.sleep(0.1)
    print("%g" % resource.getrusage(resource.RUSAGE_SELF).ru_utime)
```
And it seems to work just fine. Maybe you can try to run it and see if you get the expected result?

**Comment by CristianCantoro** [comment]
@lerks:
> I don't really know clock_gettime. Why did you choose to use that over clock[1]?

«On POSIX-compatible systems, `clock_gettime` with clock id `CLOCK_PROCESS_CPUTIME_ID` offers better resolution.» ([source][1])

I interpreted that as `clock_gettime` being more portable and having a better resolution, am I mistaken?

Furthermore, to me the problem is not the value returned by `clock_gettime` per se, the problem is that the `SIGXCPU` signal is raised way later then it should. The time returned by `clock_gettime` and /usr/bin/time` are saying the same thing, i.e. that the process has received the signal `SIGXCPU` after ~ 3 s of _user time_ execution, right?

[1]: http://en.cppreference.com/w/c/chrono/clock

**Comment by lerks** [comment]
Yes, indeed time shows that the program used 3s of user time. Weird.

I tried to replicate it locally but didn't manage. On my machine the program terminates as it should after exactly 2 seconds:
```
i: 0	- nmsec: 7	- program has been running for 214244 microseconds
i: 1	- nmsec: 9	- program has been running for 424765 microseconds
i: 2	- nmsec: 16	- program has been running for 636289 microseconds
i: 3	- nmsec: 19	- program has been running for 842958 microseconds
i: 4	- nmsec: 18	- program has been running for 1049612 microseconds
i: 5	- nmsec: 17	- program has been running for 1261296 microseconds
i: 6	- nmsec: 14	- program has been running for 1470824 microseconds
i: 7	- nmsec: 19	- program has been running for 1678844 microseconds
i: 8	- nmsec: 15	- program has been running for 1886177 microseconds

real	0m2.004s
user	0m2.000s
sys	0m0.001s
```
Since I can't reproduce it I don't really know how to help you debug it. Is there anything particular about your setup that comes to your mind that could help understand this?

**Comment by andreyv** [comment]
> > The signal flag should have type [`volatile std::sig_atomic_t`](http://en.cppreference.com/w/cpp/utility/program/sig_atomic_t). The volatile part is especially important.

> Can you clarify, please?

The `helpers::timeout_flag` variable should have the type mentioned. Interrupt handlers run asynchronously to the main program, so care must be taken. Without `std::sig_atomic_t`, access to the variable may not be atomic. Without `volatile`, the main program may not see changes to the flag variable because it is cached or optimized.

**Comment by CristianCantoro** [comment]
@lerks said:
> I tried to replicate it locally but didn't manage. On my machine the program terminates as it should after exactly 2 seconds

I am in the same situation, on my laptop it terminates after exactly 2 seconds, if I run it on a virtual machine with on the same laptop I have the problem above.

**Comment by CristianCantoro** [comment]
I have found this information:
«The CPU limit is a limit on CPU seconds rather than elapsed time. CPU seconds is basically how many seconds the CPU has been in use and does not necessarily directly relate to the elapsed time.»
([source][1])

This could be a problem because I think that the time spent for I/O is counted by CMS but not by `setrlimit()`.

However, this does not explain completely why the signal gets raised later, that's a mystery for me as of now.

I've also found that maybe using [`setitimer`][2] is the way to go. Some feedback? @andreyv, what do you think about it?

[1]: https://stackoverflow.com/questions/11661020/setrlimit-isnt-reliable
[2]: https://stackoverflow.com/questions/30007769/how-to-accurately-kill-child-process-after-certain-time-limit-in-c

**Comment by CristianCantoro** [comment]
> I've also found that maybe using setitimer is the way to go.

Here's another [gist][1], using `setitimer()`. It seems it has the same problem as the one using `setrlimit()`, i.e the signal gets raised later and not entirely predictably.

[1]: https://gist.github.com/CristianCantoro/c7a8cc7f03c6ccb7e270b45d9f6da712

**Comment by lerks** [comment]
I'm sorry no one was able to help you out on this. I personally don't know nearly enough on virtual machines to know what to look for. Maybe the way the kernel measures user time relies on processor features that aren't correctly or efficiently virtualized?
One issue I was made aware of (by @wil93 I think) that affects containers (or anything that uses a PID namespace) is that when the program runs with PID 1 its behavior in terms of signal reception is different. Maybe this could be relevant in this scenario as well?

**Comment by CristianCantoro** [comment]
I have added a question on Stack Overflow about this strange behavior: https://stackoverflow.com/questions/49033151/sigxcpu-raised-by-setrlimit-rlimit-cpu-later-than-expected-in-a-virtual-machine

**Comment by CristianCantoro** [comment]
@lerks said:
>  One issue I was made aware of (by @wil93 I think) that affects containers (or anything that uses a PID namespace) is that when the program runs with PID 1 its behavior in terms of signal reception is different. Maybe this could be relevant in this scenario as well?

I don't see how, I am using Vagrant+VirtualBox, I am not using containers.

**Comment by andreyv** [comment]
I don't think `setitimer()` is anything special, `setrlimit()` with `RLIMIT_CPU` should work too. The following program works for me on a real computer and in an Ubuntu VM running in QEMU. You can try it too.

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <sys/resource.h>

static volatile sig_atomic_t flag;

static void handler(int sig)
{
    flag = sig;
}

int main()
{
    struct rlimit rl = { .rlim_cur = 2, .rlim_max = 10 };

    if (setrlimit(RLIMIT_CPU, &rl) < 0)
    {
        perror("setrlimit");
        return EXIT_FAILURE;
    }

    struct sigaction sa = { .sa_handler = handler };

    if (sigaction(SIGXCPU, &sa, NULL) < 0)
    {
        perror("sigaction");
        return EXIT_FAILURE;
    }

    while (!flag);

    printf("Received %s after %f s\n", strsignal(flag), (double)clock() / CLOCKS_PER_SEC);
}
```

You could check whether hardware acceleration is enabled in your VM and whether the VM is running in paravirtualized mode.

**Comment by CristianCantoro** [comment]
Hi @andreyv, if I compile your code I get this error:
```
$ /usr/bin/g++ -Wall -O2 -pipe -static -s -o test test.cpp
test.cpp: In function ‘int main()’:
test.cpp:25:29: error: expected primary-expression before ‘.’ token
     struct sigaction sa = { .sa_handler = handler };
```

I will take a look to the virtualization setting of my instance of VirtualBox.

**Comment by andreyv** [comment]
It's C code, compile it with `gcc`.

**Comment by CristianCantoro** [comment]
I have the same problem in the virtual machine, see the results
```
$ for i in $(seq -w 1 10); do echo -en "$i\t - "; ./test_andreyv; done
01	 - Received CPU time limit exceeded after 3.275391 s
02	 - Received CPU time limit exceeded after 3.432943 s
03	 - Received CPU time limit exceeded after 3.660162 s
04	 - Received CPU time limit exceeded after 3.666657 s
05	 - Received CPU time limit exceeded after 3.431784 s
06	 - Received CPU time limit exceeded after 3.484979 s
07	 - Received CPU time limit exceeded after 3.220757 s
08	 - Received CPU time limit exceeded after 3.712474 s
09	 - Received CPU time limit exceeded after 3.352505 s
10	 - Received CPU time limit exceeded after 3.695584 s
```
Luckily, it seems that I don't have this problem on the instance (which is ritualized as well, but I imagine with another system) where I have CMS installed in production.

Also within CMS everything is behaving normally in production.

So, probably this is VirtualBox problem and I think that the code using `getrlimit()/setrlimit()` ([gist](https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d)) is good to go.

**Comment by CristianCantoro** [comment]
Updated [gist](https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d).

I am happy to report that we have been using this setup with success for lab exercises and we are using it in a contest we are running.

**Comment by CristianCantoro** [comment]
I can also provide more feedback on how and why we are using this setup.

@lerks said:
> Having the signal handler (installed by the admin-provided grader) kill the
program is no different (from the contestant's code's perspective) than CMS
killing the evaluation when the time limit is exceeded. Hence I really
don't see the need to have that. As Andrey said all could just be based off
the clock instead.

The main reason for which we need the soft signal is that if we wait until the hard timeout CMS will give a score of 0 to the submission. Instead, we wanted a system were approximate solution could be calculated and written in the output with the idea that with more time contestants should be able to get a better approximation.

In the end, we opted for the solution of having the contestants write their solution terminated by a `#` character, one solution per line. The `checker` considers in the output file the last line terminated by a `#` as the valid output for that submission.

We also like this solution because the overhead for the contestants in minimal, they just need to include the header file and as first instruction in the main call `helpers::setup()`
```
#include "taskname.h"
int  main() {
helpers :: setup();
...
}
```

**Comment by lerks** [comment]
Thanks for the update, good to know that in the end you managed to get everything to run as you wanted! I think this shows that CMS is able to handle situations like these without changes to the code, and the comments in this issue show many different approaches that can be adopted to achieve this. It turns out there's no need for what this issue originally set out to do, so I'm closing it.

**Comment by CristianCantoro** [comment]
One year later, we are still perfecting the art of using CMS to assign tasks based on approximation algorithms.

### Recap

The basic idea is that we want to use CMS to assign tasks where contestants need to come up with approximate solutions to hard computational problems, such the _"travel salesman problem"_. The algorithms submitted should be able to calculate approximate solutions, writing them to the output as they are found. With more time, they should be able to get better approximations.

To this end, we provide custom [header files](https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d), that are also used by CMS when compiling the submissions, that define a set of helper functions. These functions set up a time limit with soft signals, which is smaller than the hard time limit define in the task settings. They also define and set the appropriate signal handlers. When the (soft) time limit is reached, the programs exits immediately (using [`quick_exit()`](http://www.cplusplus.com/reference/cstdlib/quick_exit/)), so that the submission is valid and correcly evaluate by CMS.

For this solution to work, we have to require that all submissions call the `helper::setup()` function as soon as the main starts, like this:
```cpp
#include "taskname.h"
int main() {
helpers::setup();
...
}
```

### New solution

Now, we have found a solution to eliminate this requirement, see the usual [gist](https://gist.github.com/CristianCantoro/8b729ac341fb38caaa38f4676c7ec12d).

Basically, the new header file, which need to be named `taskname.h`, wraps the existing `main` so that `helpers::setup()` is called before calling the original `main`.

```cpp
int main() {
	helpers::setup();
	return __real_main();
};
```
where `__real_main` is the original `main` function.

So, as of now, we only have two requirements for the system to work:
* include the header file `taskname.h` in the submission;
* declare the `main` function as `int main()` or `int main(void)`, any other signature will cause a compilation error.

```cpp
#include "taskname.h"
int main() {
...
}
```

We will be testing this new solution in the next weeks, we hope that this modification make the system even less intrusive and easier to use.
